infrastructures/infrastructure-factory/src/main/java/org/eclipse/che/api/factory/server/scm/kubernetes/KubernetesPersonalAccessTokenManager.java

```diff
@@ -0,0 +1,178 @@
+
+  private final KubernetesNamespaceFactory namespaceFactory;
+  private final KubernetesClientFactory clientFactory;
+  private final ScmPersonalAccessTokenFetcher scmPersonalAccessTokenFetcher;
+
+  @Inject
+  public KubernetesPersonalAccessTokenManager(
+      KubernetesNamespaceFactory namespaceFactory,
+      KubernetesClientFactory clientFactory,
+      ScmPersonalAccessTokenFetcher scmPersonalAccessTokenFetcher) {
+    this.namespaceFactory = namespaceFactory;
+    this.clientFactory = clientFactory;
+    this.scmPersonalAccessTokenFetcher = scmPersonalAccessTokenFetcher;
+  }
+
+  @VisibleForTesting
+  void save(PersonalAccessToken personalAccessToken)
+      throws UnsatisfiedPreconditionException, ScmConfigurationPersistenceException {
+    try {
+      String namespace = getFirstNamespace();
+      ObjectMeta meta =
+          new ObjectMetaBuilder()
+              .withName(NameGenerator.generate(NAME_PATTERN, 5))
+              .withAnnotations(
+                  new ImmutableMap.Builder<String, String>()
+                      .put(ANNOTATION_CHE_USERID, personalAccessToken.getCheUserId())
+                      .put(ANNOTATION_SCM_USERID, personalAccessToken.getScmUserId())
+                      .put(ANNOTATION_SCM_USERNAME, personalAccessToken.getScmUserName())
+                      .put(ANNOTATION_SCM_URL, personalAccessToken.getScmProviderUrl())
+                      .put(
+                          ANNOTATION_SCM_PERSONAL_ACCESS_TOKEN_ID,
+                          personalAccessToken.getScmTokenId())
+                      .put(
+                          ANNOTATION_SCM_PERSONAL_ACCESS_TOKEN_NAME,
+                          personalAccessToken.getScmTokenName())
+                      .build())
+              .withLabels(SECRET_LABELS)
+              .build();
+
+      Secret secret =
+          new SecretBuilder()
+              .withApiVersion("v1")
+              .withMetadata(meta)
+              .withData(
+                  Map.of(
+                      "token",
+                      Base64.getEncoder()
+                          .encodeToString(
+                              personalAccessToken.getToken().getBytes(StandardCharsets.UTF_8))))
+              .build();
+
+      clientFactory.create().secrets().inNamespace(namespace).createOrReplace(secret);
+    } catch (KubernetesClientException | InfrastructureException e) {
+      throw new ScmConfigurationPersistenceException(e.getMessage(), e);
+    }
+  }
+
+  @Override
+  public PersonalAccessToken fetchAndSave(String cheUserId, String scmServerUrl)
+      throws UnsatisfiedPreconditionException, ScmConfigurationPersistenceException,
+          ScmUnauthorizedException, ScmCommunicationException, UnknownScmProviderException {
+    PersonalAccessToken personalAccessToken =
+        scmPersonalAccessTokenFetcher.fetchPersonalAccessToken(cheUserId, scmServerUrl);
+    save(personalAccessToken);
```

Comment:
can `personalAccessToken` be null?

Reply:
yes if this concrete `PersonalAccessTokenFetcher` implementation can't fech it for the given url. 
I should better document that - true.